File: day4/part2.cpp
#include <bits/stdc++.h>
using namespace std;

ifstream in("in.txt");
ofstream out("out.txt");

int main() {
    string line;
    int sum = 0;
    vector<vector<int>> arr {
        {1, 1}, {1, -1}, {-1, -1}, {-1, 1},
        {1, 0}, {0, 1}, {-1, 0}, {0, -1}
    };
    
    vector<string> tmp;
    while (in >> line) {
        tmp.push_back(line);
    }

    while(true){
        vector<vector<int>> ans;
        for (int i = 0; i < tmp.size(); i++) {
            for (int j = 0; j < tmp[i].size(); j++) {
                int count = 0;
    
                for (int k = 0; k < arr.size(); k++) {
                    int ni = i + arr[k][0];
                    int nj = j + arr[k][1];
    
                    if (ni >= 0 && ni < tmp.size() && nj >= 0 && nj < tmp[i].size()) {
                        if (tmp[ni][nj] == '@') count++;
                    }
                }
    
                if (tmp[i][j] == '@' && count <= 3) {
                    ans.push_back({ i, j });
                }
            }
        }
        
        for (int i = 0; i < ans.size(); i++) {
            tmp[ans[i][0]][ans[i][1]] = 'X'; 
        }
        
        if(ans.size() == 0) break;
        sum = sum + ans.size();
    }
    

    out << sum;
}

File: day6/part2.py
import re
from functools import reduce
import operator

def calc(tmp, symbol):
    result = 0 
    if symbol == "+":
        result = sum(tmp)

    elif symbol == "*":
        result = 1
        for x in tmp:
            result *= x
    return result 
    
with open('in.txt', 'r') as file:
    ans = 0
    mxl = 0
    arr = []
    for line in file:
        arr.append(line.replace("\n", ""))
        mxl = max(mxl, len(line.replace("\n", "")))
    print(arr)
    tmp = []
    symbol = ""
    for i in range(mxl):
        if len(arr[-1])>i and arr[-1][i] != ' ':
            print(symbol, tmp)
            ans = ans + calc(tmp, symbol)
            symbol = arr[-1][i]
            tmp = []
        num = 0
        for j in range(len(arr) - 1):
            if arr[j][i] == ' ':
                continue
            num = num * 10 + int(arr[j][i])
        if num != 0:
            tmp.append(num)
    ans = ans + calc(tmp, symbol)
    print(symbol, tmp)
    print(ans)